validator-network/src/network_impl.rs

```diff
@@ -43,37 +45,30 @@ where
     pub fn new(network: Arc<N>) -> Self {
         Self {
             network,
-            state: RwLock::new(State {
+            state: Arc::new(RwLock::new(State {
+                own_validator_id: None,
                 validator_keys: vec![],
                 validator_peer_id_cache: BTreeMap::new(),
-            }),
+            })),
         }
     }
 
-    async fn dial_peer(&self, peer_id: N::PeerId) -> Result<(), NetworkError<N::Error>> {
-        let mut event_stream = self.network.subscribe_events();
-
-        if self.network.has_peer(peer_id) {
-            return Ok(());
+    /// For use in closures, so that no reference to `self` needs to be kept around.
+    fn arc_clone(&self) -> ValidatorNetworkImpl<N> {
+        ValidatorNetworkImpl {
+            network: Arc::clone(&self.network),
+            state: Arc::clone(&self.state),
         }
+    }
 
-        self.network.dial_peer(peer_id).await?;
-
-        let future = async move {
-            loop {
-                match event_stream.next().await {
-                    Some(Ok(NetworkEvent::PeerJoined(joined_id, _))) if joined_id == peer_id => {
-                        break Ok(())
-                    }
-                    Some(Err(_)) | None => break Err(NetworkError::Offline), // TODO Error type?
-                    _ => {}
-                }
-            }
-        };
-
-        tokio::time::timeout(Duration::from_secs(5), future)
-            .await
-            .map_err(|_| NetworkError::Unreachable)?
+    /// Returns the local validator ID.
+    fn local_validator_id(&self) -> u16 {
```

Comment:
This will panic for non elected, but active validators.

Reply:
Made it a `Result` instead.